### PR TITLE
Add MVP LZ4 component

### PR DESCRIPTION
**Pull request by whyitfor**
- [x] I have reviewed the [OFRAK contributor guide](https://ofrak.com/docs/contributor-guide/getting-started.html) and attest that this pull request is in accordance with it.
- [x] I have made or updated a changelog entry for the changes in this pull request.

**One sentence summary of this PR (This should go in the CHANGELOG!)**
Add LZ4 compression format unpackers and packers with support for all frame types (modern, legacy, skippable)

**Link to Related Issue(s)**
N/A.

**Please describe the changes in your request.**
Lz4 Components.

Lz4Unpacker currently supports unpacking modern LZ4 format (Lz4ModernData),
legacy format (see Lz4LegacyData), and skippable data (Lz4SkippableData).

Lz4Packer supports repacking the modern LZ4 format (Lz4ModernData), matching block/checksum
information extracted during unpacking. Compression level can be specified via config.

Lz4LegacyPacker supports repacking legacy LZ4 format (Lz4LegacyData) with compression level
support (default/fast/high modes). Compression level can be specified via config.

**Anyone you think should look at this, specifically?**
@rbs-jacob

### CODE REVIEW

File: ofrak_core/src/ofrak/core/lz4.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+class Lz4Unpacker(Unpacker[None]):
+    """
+    Unpack (decompress) an LZ4 file.
+
+    Supports all LZ4 frame formats:
+    - LZ4 default frame (modern format with metadata)
+    - LZ4 legacy frame (older format for backward compatibility)
+    - LZ4 skippable frames (metadata containers)
+    """
```
Comment (whyitfor) [inline ofrak_core/src/ofrak/core/lz4.py]: Details about type of LZ4 compression need to be captured during unpacking and stored as attributes.

File: ofrak_core/src/ofrak/core/lz4.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        """
+        Pack data into LZ4 format.
+
+        :param resource: The LZ4 resource to pack
+        :param config: Optional packer configuration
+        """
+        lz4_child = await resource.get_only_child()
+        child_data = await lz4_child.get_data()
+
+        lz4_compressed = lz4.frame.compress(child_data)
```
Comment (whyitfor) [inline ofrak_core/src/ofrak/core/lz4.py]: When packing, the correct type needs to be used to maintain idemponency of unpack-pack

File: ofrak_core/src/ofrak/core/lz4.py (hunk 1 above)
Comment (whyitfor) [reply ofrak_core/src/ofrak/core/lz4.py]: Alternatively, is this information that can be gathered as part of identification?

File: ofrak_core/src/ofrak/core/lz4.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        # Check for legacy frame
+        if data == LZ4_LEGACY_MAGIC:
+            resource.add_tag(Lz4LegacyData)
+            return
+
+        # Check for skippable frames
+        # Format: 0x5X 0x2A 0x4D 0x18 where X is 0-F
+        if data[1:4] == b"\x2a\x4d\x18" and 0x50 <= data[0] <= 0x5F:
+            resource.add_tag(Lz4SkippableData)
+            return
```
Comment (rbs-jacob) [inline ofrak_core/src/ofrak/core/lz4.py]: It registers magic mime identifiers, so we don't also need this identifier.

File: ofrak_core/src/ofrak/core/lz4.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+class Lz4SkippableData(Lz4Data):
+    """
+    LZ4 skippable frame.
+
+    Special frame type for embedding metadata or application-specific data:
+    - Not compressed data
+    - Contains arbitrary bytes
+    - LZ4 parsers can safely skip these frames
+    - Typically used alongside regular frames
+    """
```
Comment (rbs-jacob) [inline ofrak_core/src/ofrak/core/lz4.py:75]: Having these other types isn't actually helpful. We just want one tag for regular `Lz4Data`.

File: ofrak_core/tests/components/test_lz4_component.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+    Test the basic unpack, modify, and pack functionality for a simple LZ4 file.
+
+    This test verifies that:
+    - A standard LZ4 file (modern frame format) can be unpacked
+    - The unpacked data can be modified
+    - The modified data can be repacked into a valid LZ4 file
+    """
+
+    def write_lz4(self, lz4_path: Path):
+        compressed_data = lz4.frame.compress(self.INITIAL_DATA)
```
Comment (rbs-jacob) [inline ofrak_core/tests/components/test_lz4_component.py]: We probably want to have static test files instead of doing compression on the fly. That way we're not testing the LZ4 library against itself, but rather testing it against real LZ4 instances found in the wild.

File: ofrak_core/CHANGELOG.md [hunk 6]
```diff
@@ -8,6 +8,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 ### Added
 - Add OFRAK requirements, requirement to test mapping, test specifications ([#656](https://github.com/redballoonsecurity/ofrak/pull/656))
 - Add `-V, --version` flag to ofrak cli ([#652](https://github.com/redballoonsecurity/ofrak/pull/652))
+- Add LZ4 compression format unpacker and packer with support for all frame types (modern, legacy, skippable) ([#PLACEHOLDER](https://github.com/redballoonsecurity/ofrak/pull/PLACEHOLDER))
```
Comment (rbs-jacob) [inline ofrak_core/CHANGELOG.md]:
```suggestion
- Add LZ4 compression format unpacker and packer with support for all frame types (modern, legacy, skippable) ([#661](https://github.com/redballoonsecurity/ofrak/pull/661))
```

File: ofrak_core/requirements.txt [hunk 7]
```diff
@@ -11,6 +11,7 @@ intervaltree==3.1.0
 keystone-engine==0.9.2
 jefferson==0.4.5;python_version>="3.8"
 lief==0.16.1
+lz4==4.4.4
```
Comment (rbs-jacob) [inline ofrak_core/requirements.txt:14]: LLMs often have out-of-date info. Did you confirm that this is actually the latest version available on PyPI?

File: ofrak_core/tests/components/test_lz4_component.py (hunk 5 above)
Comment (whyitfor) [reply ofrak_core/tests/components/test_lz4_component.py]: Yup, this was the next step

File: ofrak_core/requirements.txt (hunk 7 above)
Comment (whyitfor) [reply ofrak_core/requirements.txt:14]: Yes, and it is: https://pypi.org/project/lz4/. (This could have been an easy check before writing this comment).

File: ofrak_core/src/ofrak/core/lz4.py (hunk 3 above)
Comment (whyitfor) [reply ofrak_core/src/ofrak/core/lz4.py]: We actually do need these -- without them images are not tagged correctly. The Magic mime ones can probably be removed.

Comment (rbs-jacob) [reply ofrak_core/src/ofrak/core/lz4.py]: If we keep these magic strings, we should use your faster version of identify for byte strings.

File: ofrak_core/CHANGELOG.md [hunk 8]
```diff
@@ -8,6 +8,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 ### Added
 - Add OFRAK requirements, requirement to test mapping, test specifications ([#656](https://github.com/redballoonsecurity/ofrak/pull/656))
 - Add `-V, --version` flag to ofrak cli ([#652](https://github.com/redballoonsecurity/ofrak/pull/652))
+- Add LZ4 compression format unpacker and packer with support for all frame types (modern, legacy, skippable) ([#661](https://github.com/redballoonsecurity/ofrak/pull/661))
```
Comment (whyitfor) [inline ofrak_core/CHANGELOG.md]:
```suggestion
- Add LZ4 compression format unpackers and packers with support for all frame types (modern, legacy, skippable) ([#661](https://github.com/redballoonsecurity/ofrak/pull/661))
```

File: ofrak_core/src/ofrak/core/lz4.py (hunk 1 above)
Comment (whyitfor) [reply ofrak_core/src/ofrak/core/lz4.py]: Added this in.

File: ofrak_core/src/ofrak/core/lz4.py (hunk 3 above)
Comment (whyitfor) [reply ofrak_core/src/ofrak/core/lz4.py]: yup, done.

File: ofrak_core/tests/components/test_lz4_component.py (hunk 5 above)
Comment (whyitfor) [reply ofrak_core/tests/components/test_lz4_component.py]: I've left some corner case tests but added tests with a bunch of file formats.

File: ofrak_core/src/ofrak/core/lz4.py (hunk 4 above)
Comment (whyitfor) [reply ofrak_core/src/ofrak/core/lz4.py:75]: It is useful to treat these as three distinct types, as the code suggests.

**Comment by rbs-afflitto** [comment]
I'm attempting to unpack an LZ4 (legacy) compressed kernel with this unpacker, but I'm getting a runtime error:

```
Invalid LZ4 legacy format: header says 4623489 bytes but found 17143553
```

Using the standard lz4 tool (v1.9.4) the image unpacks. I wonder if this is a bug in the Python lz4 module.

`17143561` seems to be the file size of the lz4 file.

File: ofrak_core/tests/components/test_lz4_component.py [hunk 9]
```diff
@@ -0,0 +1,593 @@
...
+    # Verify the repacked data by unpacking it again
+    repacked_data = await resource.get_data()
+    verify_resource = await ofrak_context.create_root_resource(
+        "repacked_test.lz4", data=repacked_data
+    )
+    await verify_resource.unpack()
+
+    verify_child = await verify_resource.get_only_child()
+    verified_data = await verify_child.get_data()
+    assert verified_data.startswith(modification)
```
Comment (rbs-jacob) [inline ofrak_core/tests/components/test_lz4_component.py]:
```suggestion
    resource = await ofrak_context.create_root_resource_from_file(test_case.test_file)
    await resource.unpack()
    assert resource.has_tag(Lz4Data)
    
    initial_data = test_case.input_file.read_bytes()
    child = await resource.get_only_child()
    child_data = await child.get_data()
    assert child_data == initial_data

    modification = b"OFRAK"
    child.queue_patch(Range.from_size(0, len(modification)), modification)
    await child.save()
    await resource.pack()

    repacked_data = await resource.get_data()
    verify_resource = await ofrak_context.create_root_resource(
        "repacked_test.lz4", data=repacked_data
    )
    await verify_resource.unpack()

    verify_child = await verify_resource.get_only_child()
    verified_data = await verify_child.get_data()
    assert verified_data.startswith(modification)
```

In my opinion, these comments just add noise. This code is pretty self-explanatory, and doesn't need them. Explaining that `resource.pack` means "pack it back" is actively harmful to code readability in my opinion.

I only bring this up to get ahead of it, since I imagine there will be increasing LLM assistance when making OFRAK contributions. We will want to agree on whether comments like these are appropriate or not in order to more effectively use LLMs, and ensure they don't delay code review.

Not sure if there is a Claude setting to get it to chill out with the comments? I've heard they help it "think through" writing code, and do a better job. If that's the case, maybe an additional step at the end to go through and automatically clean up/remove many of them, specified in a Claude rules file somewhere?

There is a direct suggestion in this comment as an example of how much more concise the code could be, with added benefit of logical grouping for the expressions. But the general comment applies to most of these test functions.

I'd love to find a way to solve this and/or add it to the contributor guidelines so I don't have to point it out in nitpicky comments like this one!

File: ofrak_core/tests/components/test_lz4_component.py [hunk 10]
```diff
@@ -0,0 +1,593 @@
+"""
+Test the functionality of the LZ4 component, including unpacking,
+modifying, and repacking LZ4-compressed data.
```
Comment (rbs-jacob) [inline ofrak_core/tests/components/test_lz4_component.py:3]: I don't have issues with the tests in this file, but I'm curious why you didn't use the unpack modify pack test pattern (or other test patterns)?

File: ofrak_core/src/ofrak/core/lz4.py [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+        decompressed_data = b""
+        while deserializer.position() < len(resource_data):
+            # Legacy LZ4 has a repeating pattern of 4 bytes block size followed by compressed data
+            block_size = deserializer.unpack_uint()
+            compressed_block = deserializer.read(block_size)
+            try:
+                # LZ4 legacy block size (uncompressed) is 8 MB (see https://github.com/lz4/lz4/blob/67a385a170d2dc331a25677e0d20d96eef0450c5/programs/lz4io.c#L86)
+                decompressed_data += lz4.block.decompress(
+                    compressed_block,
+                    uncompressed_size=8 * (1 << 20),
```
Comment (rbs-jacob) [inline ofrak_core/src/ofrak/core/lz4.py:164]:
```suggestion
                    uncompressed_size=8 * 1024 * 1024,
```

Sometimes I find it clearer to express MB in terms of KB squared. Feel free to ignore this – just a matter of taste.

File: ofrak_core/src/ofrak/core/lz4.py [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+
+class Lz4Packer(Packer[Lz4PackerConfig]):
+    """
+    Pack data into a compressed LZ4 file using modern frame format.
+
+    Implementation repacks modern frame format preserving frame metadata.
+    Compression level can be specified via config (default: 0).
+    """
+
+    targets = (Lz4ModernData,)
```
Comment (rbs-jacob) [inline ofrak_core/src/ofrak/core/lz4.py:199]: Is it intentional that none of these target and pack the skippable data?

File: ofrak_core/src/ofrak/core/lz4.py [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+        :param config: Optional configuration specifying compression level
+        """
+        if config is None:
+            config = Lz4PackerConfig()
+
+        lz4_child = await resource.get_only_child()
+        child_data = await lz4_child.get_data()
+
+        # LZ4 legacy format uses 8 MB blocks (see https://github.com/lz4/lz4/blob/67a385a170d2dc331a25677e0d20d96eef0450c5/programs/lz4io.c#L86)
+        LEGACY_BLOCK_SIZE = 8 * (1 << 20)  # 8 MB
```
Comment (rbs-jacob) [inline ofrak_core/src/ofrak/core/lz4.py:262]:
```suggestion
        LEGACY_BLOCK_SIZE = 8 * 1024 * 1024  # 8 MB
```

Same comment as above about using KB squared. Once again, feel free to ignore.

File: ofrak_core/src/ofrak/core/lz4.py [hunk 14]
```diff
@@ -0,0 +1,332 @@
...
+                compressed_block = lz4.block.compress(
+                    block_data,
+                    mode="high_compression",
+                    compression=config.compression_level,
+                    store_size=False,
+                )
+
+            # Append block size + compressed block data
+            compressed_block_size = len(compressed_block)
+            lz4_compressed += compressed_block_size.to_bytes(4, "little") + compressed_block
```
Comment (rbs-jacob) [inline ofrak_core/src/ofrak/core/lz4.py]: Have you tried running this for moderately large files? Does it run slowly? I've had speedups from appending byte strings to a list, and then concatenating the list in one go. Curious if that would apply here.

**Review by rbs-jacob — CHANGES_REQUESTED** [review]
Most things fairly minor, and several are completely optional.

File: ofrak_core/src/ofrak/core/lz4.py (hunk 14 above)
Comment (whyitfor) [reply ofrak_core/src/ofrak/core/lz4.py]: I agree. @rbs-afflitto I'll fix this for you.

File: ofrak_core/src/ofrak/core/lz4.py (hunk 12 above)
Comment (whyitfor) [reply ofrak_core/src/ofrak/core/lz4.py:199]: Yes. There doesn't seem to be a good use case for now to actually pack skippable data -- once we need this we can add it in.

File: ofrak_core/src/ofrak/core/lz4.py (hunk 11 above)
Comment (rbs-afflitto) [reply ofrak_core/src/ofrak/core/lz4.py:164]: For context I chose to use the same syntax that the `lz4` repo used here. I agree it's less intuitive, but wanted to ensure it matches the reference implementation.

File: ofrak_core/tests/components/test_lz4_component.py (hunk 10 above)
Comment (whyitfor) [reply ofrak_core/tests/components/test_lz4_component.py:3]: I did originally, but we don't have a good pattern that tests that files repack to the same bytes. These also felt more "self-documenting" in a way the original wasn't.

File: ofrak_core/tests/components/test_lz4_component.py (hunk 9 above)
Comment (whyitfor) [reply ofrak_core/tests/components/test_lz4_component.py]: Worth thinking about. Sometimes its nice it tests to have comments saying what it is trying to test, this was a bit heavy-handed though.
